lib/view_component/base.rb

```diff
@@ -135,6 +146,60 @@ def with(area, content = nil, &block)
       nil
     end
 
+    # Build a Slot on a component.
+    #
+    # slot: Name of Slot, in symbol form
+    # **args: Arguments to be passed to Slot initializer
+    #
+    # For example:
+    # <%= render(SlotsComponent.new) do |component| %>
+    #   <% component.slot(:footer, class_names: "footer-class") do %>
+    #     <p>This is my footer!</p>
+    #   <% end %>
+    # <% end %>
+    #
+    def slot(slot, **args, &block)
+      # Raise ArgumentError if `slot` does not exist
+      unless slot_names.include?(slot)
+        raise ArgumentError.new "Unknown slot '#{slot}' - expected one of '#{slot_names}'"
+      end
+
+      slot_class = "#{self.class.name}::#{slot.to_s.capitalize}".constantize
+
+      # Instantiate Slot class, accommodating Slots that don't accept arguments
+      slot_instance =
+        if args.present?
+          slot_class.new(args)
+        else
+          slot_class.new
+        end
+
+      # Capture block and assign to slot_instance#content
+      slot_instance.content = view_context.capture(&block) if block_given?
+
+      if slot_class.superclass == ViewComponent::CollectionSlot
```

Comment:
On a different topic about this block of code, I think that this late binding effect for `CollectionSlot` will be confusing for users. Presently, the default value for all unspecified slots is `nil`, but I wonder if the default value for a collection slot would be better as `[]`, whether the component consumer defined a block for it or not.

For example, the Box example in the documentation currently would raise an error if `row` isn't specified by the Box consumer, because it assumes that `rows` is an empty collection, not `nil`:

```erb
<% if rows.any? %>
```

Without specifying `row`, the component consumer will get a vague `NoMethodError: undefined method 'each' for nil:NilClass`.

One possible alternative is to improve the error message that would happen here, somehow, maybe?

```erb
<% render(BoxComponent.new) do  %>
  <!-- no rows -->
<% end %>
```

```
# current:
NoMethodError: undefined method `each' for nil:NilClass

# possible?
UnspecifiedSlotError: BoxComponent requires at least one `row` slot to be defined
```

Reply:
`[]` as the default makes sense to me.